realm/realm-library/src/androidTest/java/io/realm/RealmTests.java

```diff
@@ -1056,6 +1057,44 @@ public void compactRealm_onExternalStorage() {
         Realm.deleteRealm(config);
     }
 
+    private Pair<Long, Long> populateTestRealmAndCompactOnLaunch(CompactOnLaunchCallback compactOnLaunch) {
+        final String REALM_NAME = "test.realm";
+        RealmConfiguration realmConfig = configFactory.createConfiguration(REALM_NAME);
+        Realm realm = Realm.getInstance(realmConfig);
+        populateTestRealm(realm, 100);
+        realm.close();
+        long before = new File(realmConfig.getPath()).length();
+        realmConfig = configFactory.createConfiguration(REALM_NAME, compactOnLaunch);
+        realm = Realm.getInstance(realmConfig);
+        realm.close();
+        long after = new File(realmConfig.getPath()).length();
+        return new Pair(before, after);
+    }
+
+    @Test
+    public void compactOnLaunch_shouldCompact() throws IOException {
+        Pair<Long, Long> results = populateTestRealmAndCompactOnLaunch(new CompactOnLaunchCallback() {
+            @Override
+            public boolean shouldCompact(long totalBytes, long usedBytes) {
+                assertTrue(totalBytes > usedBytes);
```

Comment:
I think you can create 100 objects and then remove them. Those 100 deleted objects will still take space in the file until compacting. So there would be obvious results after compacting called. Then it should be always true here. (until GC is implemented in core.)